File: leetcode/dp/5. Longest Palindromic Substring/js/main.js
/*
  
  Типичная проблема dp, есть описание в файле dpProblems.js

  Подходит под случай когда у нас есть 1 строка

  Идем от длины строки от [len = 1 до len = s.length] начинаем поиск при длине len = 1


  Рассмотрим случай "ababa".
  Если мы уже знали, что «bab» - это палиндром, то очевидно, что «ababa» должна быть палиндромом,
  Поскольку две левые и правые буквы одинаковы.

*/

// Time O(N^2)
// Space O(N^2)
const longestPalindrome = s => {
  let n = s.length;

  let dp = Array(n)
    .fill(null)
    .map(() => Array(n).fill(false));

  let start = 0;
  let maxLen = 1; // min palindrome equeal 1 char

  for (let i = 0; i < n; i++) {
    dp[i][i] = true;
  }

  for (let len = 1; len < n; len++) {
    for (let i = 0; i < n - len; i++) {
      let j = i + len;

      if (s[i] == s[j]) {
        dp[i][j] = j - i == 1 ? true : dp[i + 1][j - 1];
      }

      if (dp[i][j] == true && maxLen < j - i + 1) {
        start = i;
        maxLen = j - i + 1;
      }
    }
  }

  return s.substring(start, start + maxLen);
};

// Time O(N)
// Space O(1)
const longestPalindrome_II = s => {
  let n = s.length;
  let maxLen = 1;
  let startIndex = 0;

  for (let i = 0; i < n; i++) {
    let [left1, len1] = expandAroundCenter(i, i);
    let [left2, len2] = expandAroundCenter(i, i + 1);

    if (len1 > len2) {
      if (maxLen < len1) {
        maxLen = len1;
        startIndex = left1;
      }
    } else {
      if (maxLen < len2) {
        maxLen = len2;
        startIndex = left2;
      }
    }
  }

  return s.substring(startIndex, startIndex + maxLen);

  function expandAroundCenter(l, r) {
    while (l >= 0 && r < n && s[l] == s[r]) {
      l--;
      r++;
    }
    l++;
    r--;
    return [l, r - l + 1];
  }
};

File: leetcode/dp/5. Longest Palindromic Substring/py/main.py
class Solution:
    # Time O(N^2)
    # Space O(N^2)
    def longestPalindrome(self, s):
        n = len(s)
        dp = [[False] * n for _ in range(n)]
        maxLen = 1 # Min palindrome equals 1 char
        startIndex = 0

        for i in range(n):
          dp[i][i] = True

        for length in range(1, n):
          for i in range(n - length):
            j = i + length
                        
            if s[i] == s[j]:
              dp[i][j] = True if j - i == 1 else dp[i + 1][j - 1]
            
            # if it's a palindrome
            if dp[i][j] and maxLen < j - i + 1:
                maxLen = j - i + 1
                startIndex = i
                
        return s[startIndex:startIndex + maxLen]
    
    # Time O(N^2)
    # Space O(1)
    def longestPalindrome_II(self, s):
        n = len(s)
        maxLen = 1
        startIndex = 0
        
        def expandAroundCenter(i, j):
          while i >= 0 and j < n and s[i] == s[j]:
            i -= 1
            j += 1
          i += 1
          j -= 1

          return [i, j - i + 1]

        for i in range(n):
          # Palindrome can be expanded from its center, and there are only 2n − 1 such centers.
          # You might be asking why there are 2n − 1 but not n centers? 
          
          # The reason is the center of a palindrome can be in between two letters. 
          # Such palindromes have even number of letters (such as "abba") and its center are between the two 'b's.
          
          startIndex1, len1 = expandAroundCenter(i, i)

          # Even number of letters
          startIndex2, len2 = expandAroundCenter(i, i + 1) 
          
          if len1 > len2:
            if maxLen < len1:
              maxLen = len1
              startIndex = startIndex1            
          else:
            if maxLen < len2:
              maxLen = len2
              startIndex = startIndex2

        return s[startIndex: startIndex + maxLen]
            

          

